smi/adaptor/controller.go

```diff
@@ -0,0 +1,352 @@
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		log.Infof("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		log.Error(err)
+		return true
+	}
+
+	return true
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Ts resource
+// with the current status of the resource.
+func (c *SMIController) syncHandler(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, service, err := splitTrafficSplitKey(key)
+	if err != nil {
+		log.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+	// Get the Ts resource with this namespace/name
+	ts, err := c.tsclientset.SplitV1alpha1().TrafficSplits(namespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// TS does not exit anymore
+			// Check if there is a relevant SP that was created or updated by SMI Controller
+			// and clean up its dstOverrides
+			log.Infof("trafficsplit/%s is deleted, trying to cleanup the relevant serviceprofile", name)
+			sp, err := c.spclientset.LinkerdV1alpha2().ServiceProfiles(namespace).Get(ctx, c.toFQDN(service, namespace), metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+
+			// Empty dstOverrides in the SP
+			if ignoreAnnotationPresent(sp) {
+				log.Infof("skipping clean up of serviceprofile/%s as ignore annotation is present", sp.Name)
+				return nil
+			}
+
+			sp.Spec.DstOverrides = nil
+			_, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(namespace).Update(ctx, sp, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+			log.Infof("cleaned up `dstOverrides` of serviceprofile/%s", sp.Name)
+			return nil
+		}
+		return err
+	}
+
+	// Check if the Service Profile is already present
+	sp, err := c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Get(ctx, c.toFQDN(ts.Spec.Service, ts.Namespace), metav1.GetOptions{})
+	if err != nil {
+		// Create a Service Profile resource as it does not exist
+		sp, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Create(ctx, c.toServiceProfile(ts), metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+		log.Infof("created serviceprofile/%s for trafficsplit/%s", sp.Name, ts.Name)
+	} else {
+		log.Infof("serviceprofile/%s already present", sp.Name)
+		// Check if SP Matches the TS, and update if it not
+		spFromTs := c.toServiceProfile(ts)
+		if !equal(spFromTs, sp) {
+			log.Infof("serviceprofile/%s does not match trafficscplit/%s", sp.Name, ts.Name)
+			if ignoreAnnotationPresent(sp) {
+				log.Infof("skipping updation of serviceprofile/%s as ignore annotation is present", sp.Name)
+				return nil
+			}
+			updateDstOverrides(sp, ts, c.clusterDomain)
+			_, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Update(ctx, sp, metav1.UpdateOptions{})
```

Comment:
if you get rid of the workqueue, you will need to add back in some kind of retry behavior for these calls to update.  If the resource has changed since you did the `get`, the resource version won't match and the update will be rejected.  In that case you will need to go back through the get-modify-update process again to use the most recent version of the resource